pkg/apis/core/v1alpha1/types_shootsnapshot.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ShootSnapshot contains a snapshot of the Shoot's state required to migrate the Shoot's control plane to a new Seed.
+type ShootSnapshot struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Specification of the ShootSnapshot.
+	// +optional
+	Spec ShootSnapshotSpec `json:"spec,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ShootSnapshotList is a list of ShootSnapshot objects.
+type ShootSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list object metadata.
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+	// Items is the list of ShootSnapshots.
+	Items []ShootSnapshot `json:"items"`
+}
+
+// ShootSnapshotSpec is the specification of the ShootSnapshot.
+type ShootSnapshotSpec struct {
+	// Gardenlet holds the state of resources generated, deployed and reconciled by the gardenlet
+	Gardenlet GardenletResources `json:"gardenlet"`
+	// Extensions holds the state of custom resources reconciled by extension controllers in the seed
+	// +optional
+	Extensions []ExtensionResourceState `json:"extensions,omitempty"`
+}
+
+// GardenletResources contains the state of resources which are generated, deployed and reconciled by the gardenlet.
+type GardenletResources struct {
+	// Secrets contains an array of objects which contain data from generated secrets
+	// +optional
+	Secrets []GardenletSecretData `json:"secrets,omitempty"`
+}
+
+// GardenletSecretData holds the state of a secret which has to be generated and deployed in the shoot control plane
+// by gardenlet and/or used to generate other secrets.
+type GardenletSecretData struct {
+	// Name of the resource
+	Name string `json:"name"`
+	// Data which has to be saved
+	Data map[string]string `json:"data"`
+}
```

Comment:
It was especially not intended to replicate secrets. This would imprint the structure of generated secrets on the shoot state object. Instead the semantical data, like ssh public and private key, ca cert and key, and so on should be stored.
This information is then used by the gardener's reconciliation to create dedicated objects required by the actual implementation. For example, the same information from the state might be used for several secrets. The ca is used to sign certs for any number of required kubeconfigs which are stored along with the ca certificate in multiple secrets.
It does not seem to be a good idea to store all those replicated information separately in the shoot state.

Reply:
This struct is intended to save exactly that. It would look something like:
```
secrets
- name: ca
  data:
    ca.key: "...."
    ca.crt: "..."
- name: ssh-keypair
  data:
    id_rsa: "..."
    id_rsa.pub: "...."
```

Then those would be used to generate other secrets which will not get saved in this resource.
Perhaps the field name and comment is a bit misleading?

Reply:
@mandelsoft updated the comments around those fields. Could you have another look.

Reply:
Ok, if it's meant this way, I'm fine. But indeed the name is then misleading, because a secret is a well-known object in a kubernetes cluster. For extensions the type is directly called `ExtensionsResourceState`. So, why not omit this additional intermediate type with only one field for the `Gardener` part. Just use a similar type `GardenerResourceState`  array for the `Gardener` field with the `Name`/`Kind`/`Type` field and the `Data` map.